packages/nostr/src/client/conn.ts

```diff
@@ -137,25 +139,139 @@ export const enum OutgoingKind {
  */
 export interface OutgoingEvent {
   kind: OutgoingKind.Event
-  event: Event
+  signed: SignedEvent
 }
 
 /**
- * Outgoing "REQ" message, representing a subscription.
+ * Outgoing "REQ" message, which opens a subscription.
  */
-export interface OutgoingSubscription {
+export interface OutgoingOpenSubscription {
   kind: OutgoingKind.Subscription
   id: SubscriptionId
-  filters: Filters[]
+  filters: Filters
 }
 
 /**
- * Outgoing "CLOSE" message, representing an unsubscription.
+ * Outgoing "CLOSE" message, which closes a subscription.
  */
-export interface OutgoingUnsubscription {
+export interface OutgoingCloseSubscription {
   kind: OutgoingKind.Unsubscription
   id: SubscriptionId
 }
 
 type IncomingMessageCallback = (message: IncomingMessage) => unknown
-type ConnErrorCallback = (error: ProtocolError) => unknown
+type ErrorCallback = (error: ProtocolError) => unknown
+
+interface RawFilters {
+  ids?: string[]
+  authors?: string[]
+  kinds?: number[]
+  ["#e"]?: string[]
+  ["#p"]?: string[]
+  since?: number
+  until?: number
+  limit?: number
+}
+
+async function parseIncomingMessage(data: string): Promise<IncomingMessage> {
+  const json = parseJson(data)
+  if (!(json instanceof Array)) {
+    throw new ProtocolError(`incoming message is not an array: ${data}`)
+  }
+  if (json.length === 3) {
```

Comment:
This is very definitely going to break, we should not use the length because relays and clients implement NIP's in random orders and all NIP's are optional, only the tag from `json[0]` should be used to determine what sort of message this is